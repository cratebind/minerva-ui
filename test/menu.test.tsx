import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import 'jest-styled-components';
import {
  MenuContainer,
  MenuList,
  MenuButton,
  MenuItem,
  MenuLink,
  Icon,
  ThemeProvider,
} from '../src';
// import { axe, toHaveNoViolations } from 'jest-axe';
// expect.extend(toHaveNoViolations);

const ExampleMenu = () => (
  <MenuContainer data-testid="menu-container">
    <MenuButton data-testid="menu-button">
      Actions <Icon name="chevron-down" ml={2} size="14px" />
    </MenuButton>
    <MenuList data-testid="menu-list">
      <MenuItem onSelect={() => alert('Download')}>Download</MenuItem>
      <MenuItem onSelect={() => alert('Copy')}>Create a Copy</MenuItem>
      <MenuItem onSelect={() => alert('Mark as Draft')}>Mark as Draft</MenuItem>
      <MenuItem onSelect={() => alert('Delete')}>Delete</MenuItem>
      <MenuLink href="https://minerva-ui-docs.vercel.app">
        Open Documentation
      </MenuLink>
    </MenuList>
  </MenuContainer>
);

// @TODO: Figure out how to test components being rendered in portals
describe('<Menu />', () => {
  it('should render', () => {
    const { container, queryByTestId } = render(
      <ThemeProvider>
        <ExampleMenu />
      </ThemeProvider>
    );

    expect(queryByTestId('menu-button')).toBeInTheDocument();

    expect(container).toMatchSnapshot('Closed Menu');

    fireEvent.click(screen.getByTestId('menu-button'));

    expect(container).toMatchSnapshot('Open Menu');
  });

  // it('should show <MenuList /> when clicked', () => {
  //   const { queryByTestId, baseElement } = render(
  //     <ThemeProvider>
  //       <ExampleMenu />
  //     </ThemeProvider>
  //   );

  //   // debug();
  //   // console.log(getByRole('menu'))
  //   // console.log(document.body.querySelector('[data-reach-menu=""]'))
  //   // console.log(baseElement.querySelector('[data-reach-menu=""]'))
  //   const menu = baseElement.querySelector('[data-reach-menu=""]')
  //   expect(menu && menu.getAttribute('hidden')).not.toBeNull()
  //   expect(queryByTestId('menu-list')).not.toBeInTheDocument();
  // });

  // it('should render ModalHeader when component is called and when isOpen is true', () => {
  //   const testId = 'modal-header';
  //   const modalContent = 'Menu Header';
  //   const { queryByText, queryByTestId } = render(
  //     <ThemeProvider>
  //       <Menu isOpen>
  //         <ModalHeader data-testid={testId}>{modalContent}</ModalHeader>
  //       </Menu>
  //     </ThemeProvider>
  //   );
  //   expect(queryByTestId(testId)).toBeInTheDocument();
  //   expect(queryByText(modalContent)).toBeInTheDocument();
  // });

  // it('should render Menu Body when component is called and when isOpen is true', () => {
  //   const testId = 'modal-body';
  //   const modalContent = 'Menu Body';
  //   const { queryByText, queryByTestId } = render(
  //     <ThemeProvider>
  //       <Menu isOpen>
  //         <ModalBody data-testid={testId}>{modalContent}</ModalBody>
  //       </Menu>
  //     </ThemeProvider>
  //   );
  //   expect(queryByTestId(testId)).toBeInTheDocument();
  //   expect(queryByText(modalContent)).toBeInTheDocument();
  // });

  // it('should render Menu Footer when component is called and when isOpen is true', () => {
  //   const testId = 'modal-footer';
  //   const modalContent = 'Menu Footer';
  //   const { queryByText, queryByTestId } = render(
  //     <ThemeProvider>
  //       <Menu isOpen>
  //         <ModalFooter data-testid={testId}>{modalContent}</ModalFooter>
  //       </Menu>
  //     </ThemeProvider>
  //   );
  //   expect(queryByTestId(testId)).toBeInTheDocument();
  //   expect(queryByText(modalContent)).toBeInTheDocument();
  // });

  // it('should be accessible when Open', async () => {
  //   const modalContent = 'Menu Content';
  //   const { container } = render(
  //     <ThemeProvider>
  //       <Menu isOpen>{modalContent}</Menu>
  //     </ThemeProvider>
  //   );

  //   const results = await axe(container);

  //   expect(results).toHaveNoViolations();
  // });

  // it('should be accessible when Closed', async () => {
  //   const modalContent = 'Menu Content';
  //   const { container } = render(
  //     <ThemeProvider>
  //       <Menu isOpen={false}>{modalContent}</Menu>
  //     </ThemeProvider>
  //   );

  //   const results = await axe(container);

  //   expect(results).toHaveNoViolations();
  // });
});
