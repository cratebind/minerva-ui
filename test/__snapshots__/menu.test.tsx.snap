// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render when positioned to Right: Closed Menu 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 150ms ease 0s;
  transition: all 150ms ease 0s;
  outline: none;
  cursor: pointer;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  background-color: #fff;
  border-width: 1px;
  color: #374151;
  font-weight: 500;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  font-size: 14px;
  line-height: 20px;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  border-radius: 5px;
  border-style: solid;
  border-color: #d2d6dc;
}

.c1:hover {
  background-color: #f9fafb;
}

.c1:focus {
  border-color: #a4cafe;
  box-shadow: 0 0 0 3px rgba(118,169,250,.45);
  outline: 0;
}

.c1:active,
.c1[data-active=true] {
  border-color: #a4cafe;
  box-shadow: 0 0 0 3px rgba(118,169,250,.45);
  outline: 0;
}

.c1:disabled,
.c1:disabled:focus,
.c1:disabled:hover,
.c1[aria-disabled=true],
.c1[aria-disabled=true]:focus,
.c1[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

<div>
  <button
    aria-busy="false"
    aria-controls="menu--7"
    aria-haspopup="true"
    class="c0 c1"
    data-reach-menu-button=""
    data-testid="menu-button"
    id="menu-button--menu--7"
    role="button"
    type="button"
  >
    Actions 
  </button>
</div>
`;

exports[`<Menu /> should render when positioned to Right: Open Menu 1`] = `
<div>
  <button
    aria-busy="false"
    aria-controls="menu--7"
    aria-haspopup="true"
    class="sc-bdnxRM sc-gtsrHT eJeGwH ksncAa"
    data-reach-menu-button=""
    data-testid="menu-button"
    id="menu-button--menu--7"
    role="button"
    type="button"
  >
    Actions 
  </button>
</div>
`;

exports[`<Menu /> should render: Closed Menu 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: all 150ms ease 0s;
  transition: all 150ms ease 0s;
  outline: none;
  cursor: pointer;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  background-color: #fff;
  border-width: 1px;
  color: #374151;
  font-weight: 500;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  font-size: 14px;
  line-height: 20px;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 16px;
  padding-right: 16px;
  border-radius: 5px;
  border-style: solid;
  border-color: #d2d6dc;
}

.c1:hover {
  background-color: #f9fafb;
}

.c1:focus {
  border-color: #a4cafe;
  box-shadow: 0 0 0 3px rgba(118,169,250,.45);
  outline: 0;
}

.c1:active,
.c1[data-active=true] {
  border-color: #a4cafe;
  box-shadow: 0 0 0 3px rgba(118,169,250,.45);
  outline: 0;
}

.c1:disabled,
.c1:disabled:focus,
.c1:disabled:hover,
.c1[aria-disabled=true],
.c1[aria-disabled=true]:focus,
.c1[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

<div>
  <button
    aria-busy="false"
    aria-controls="menu--1"
    aria-haspopup="true"
    class="c0 c1"
    data-reach-menu-button=""
    data-testid="menu-button"
    id="menu-button--menu--1"
    role="button"
    type="button"
  >
    Actions 
  </button>
</div>
`;

exports[`<Menu /> should render: Open Menu 1`] = `
<div>
  <button
    aria-busy="false"
    aria-controls="menu--1"
    aria-haspopup="true"
    class="sc-bdnxRM sc-gtsrHT eJeGwH ksncAa"
    data-reach-menu-button=""
    data-testid="menu-button"
    id="menu-button--menu--1"
    role="button"
    type="button"
  >
    Actions 
  </button>
</div>
`;
